Adding https_config options to the alicloud_cdn_domain resource

chanind opened the pull request:
This PR adds the ability to set a SSL certificate and enable HTTPs on the `alicloud_cdn_domain` resource. Previously it was not possible to enable SSL through terraform.

This is my first time both writing Go and working on a terraform provider, so apologies if the code is strange!

Fixes https://github.com/terraform-providers/terraform-provider-alicloud/issues/591

xiaozhu36 commented:
HI @chanind Thanks for your contribution. Can you add a testcase for this PR?

chanind commented:
@xiaozhu36 added a test case in d41d271 and fixed up some minor issues I found in the process

xiaozhu36 commented on alicloud/resource_alicloud_cdn_domain.go:
```diff
@@ -164,6 +164,32 @@ func resourceAlicloudCdnDomain() *schema.Resource {
 				MaxItems: 1,
 			},
 
+			"https_config": &schema.Schema{
```
I think `certificate_config` will be better.

xiaozhu36 commented on alicloud/resource_alicloud_cdn_domain.go:
```diff
@@ -389,6 +415,18 @@ func resourceAlicloudCdnDomainUpdate(d *schema.ResourceData, meta interface{}) e
+		if d.IsNewResource() {
+			err := WaitForDomainStatus(d.Id(), Online, 360, meta)
+			if err != nil {
+				return fmt.Errorf("Timeout when Cdn Domain Online")
```
Please print err details before return err.

chanind replied:
This behavior was copied from https://github.com/terraform-providers/terraform-provider-alicloud/blob/master/alicloud/resource_alicloud_cdn_domain.go#L295. Should that part of the provider also be changed?

xiaozhu36 replied:
yes

xiaozhu36 commented on alicloud/resource_alicloud_cdn_domain.go:
```diff
@@ -429,6 +467,22 @@ func resourceAlicloudCdnDomainRead(d *schema.ResourceData, meta interface{}) err
+			"server_certificate":        domain.ServerCertificate,
+			"server_certificate_status": domain.ServerCertificateStatus,
+		}
+		if oldConfig != nil && oldConfig.Len() > 0 {
```
I think condition `if oldConfig != nil && oldConfig.Len() > 0` is redundant. private_key is sensitive and it is no need to return.

chanind replied:
I didn't want to put `private_key` into the state, but I find that if I don't write it then terraform always says it changed. Is there a way to not write `private_key` but also not have terraform keep thinking there are new changes?

xiaozhu36 replied:
you can set private_key "Compute: true"

chanind replied:
@xiaozhu36 maybe it's easier if you change this PR so it works the way you suggest? You should be able to push changes. When I added `Compute: true` and removed `private_key` from the state terraform thinks the resource has changed every run. If I'm doing something wrong it's probably easier for you to push a fix since you're more experienced than I am.

xiaozhu36 commented on alicloud/resource_alicloud_cdn_domain.go:
```diff
@@ -716,6 +770,64 @@ func authConfigUpdate(client *connectivity.AliyunClient, d *schema.ResourceData)
+		})
+		if err != nil {
+			return err
+		}
```
Line 788 needs set partial after ser domain config.

chanind replied:
I used `authConfigUpdate` as a template, from here: https://github.com/terraform-providers/terraform-provider-alicloud/blob/master/alicloud/resource_alicloud_cdn_domain.go#L663. There is a `d.SetPartial("https_config")` on line 792. Why does one also need to be on line 788? It seems like in `authConfigUpdate` there is only 1 `SetPartial`. Sorry if this is a dumb question, I don't fully understand what `SetPartial` does.

xiaozhu36 replied:
When one config has been changed successfully by invoking API, SetPartial will write down it to state even if the next change failed.

xiaozhu36 commented on alicloud/resource_alicloud_cdn_domain.go:
```diff
@@ -716,6 +770,64 @@ func authConfigUpdate(client *connectivity.AliyunClient, d *schema.ResourceData)
+	if okServerCertificate {
+		err := WaitForHttpsServerCertificate(client, d.Id(), args.ServerCertificate, 360)
+		if err != nil {
+			return fmt.Errorf("Timeout waiting for Cdn server certificate")
```
Please print error details when returning error.

xiaozhu36 replied:
Please print error details when returning error.

xiaozhu36 commented on alicloud/resource_alicloud_cdn_domain_test.go:
```diff
@@ -112,6 +112,36 @@ func TestAccAlicloudCdnDomain_basic(t *testing.T) {
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.#", "1"),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.3358202545.server_certificate_status", "on"),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.3358202545.server_certificate", strings.Replace(testServerCertificate, `\n`, "\n", -1)),
+					resource.TestCheckResourceAttr("alicloud_cdn_domain.domain", "https_config.3358202545.private_key", strings.Replace(testPrivateKey, `\n`, "\n", -1)),
```
I think the hash code should not occur in testcase because of it always changes. If you want to check `server_certificate_status `  and other config, you can consider to modify https_config type to TypeList.

chanind replied:
I saw this pattern in the AWS provider, ex https://github.com/terraform-providers/terraform-provider-aws/blob/5a062fcdc5b24d82017aa15bc561fac1f152a2c6/aws/resource_aws_elb_test.go#L122. I agree it's not ideal to include the hash code in the test, but it seems like we shouldn't need to change the provider structure just to avoid dealing with hash codes in the state. The other params of `alicloud_cdn_domain` all use `schema.TypeSet`, ex `auth_config`, `refer_config`, `page_404_config`, and `parameter_filter_config` so I wanted to be consistent. Is there another helper I can use to write assertions that work on `schema.TypeSet` values, or is it not possible to test `schema.TypeSet` types?

xiaozhu36 replied:
If you can ensure same config has same hash code, the testcase can be kept. Can you run the testcase and upload the result when next submitting?

chanind replied:
The hash shouldn't change unless we change the values contained in the config. Here is the output of the test:
```
=== RUN   TestAccAlicloudCdnDomain_https
2019/01/04 11:32:07 [WARN] Test: Executing step 0
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypeValidate
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypePlan
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:32:07 [WARN] Test: Step plan: DIFF:

CREATE: alicloud_cdn_domain.domain
  cdn_type:                                                "" => "web"
  certificate_config.#:                                    "0" => "1"
  certificate_config.3358202545.private_key:               "<sensitive>" => "<sensitive>" (attribute changed)
  certificate_config.3358202545.server_certificate:        "<sensitive>" => "<sensitive>" (attribute changed)
  certificate_config.3358202545.server_certificate_status: "" => "on"
  domain_name:                                             "" => "tf-testacc6068554464636384429.xiaozhu.com"
  scope:                                                   "" => "overseas"
  source_port:                                             "" => "80"
  source_type:                                             "" => "oss"
  sources.#:                                               "0" => "1"
  sources.21942838:                                        "" => "terraformtest.aliyuncs.com"

STATE:

<no state>
2019/01/04 11:32:07 [INFO] terraform: building graph: GraphTypeApply
2019/01/04 11:32:07 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:48 [WARN] Domain id "tf-testacc6068554464636384429.xiaozhu.com"
2019/01/04 11:38:48 [INFO] terraform: building graph: GraphTypePlan
2019/01/04 11:38:48 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:48 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:38:48 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:51 [INFO] terraform: building graph: GraphTypePlan
2019/01/04 11:38:51 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:51 [WARN] Test: Running ID-only refresh check on tf-testacc6068554464636384429.xiaozhu.com
2019/01/04 11:38:51 [WARN] Test: Executing destroy step
2019/01/04 11:38:51 [INFO] terraform: building graph: GraphTypeValidate
2019/01/04 11:38:51 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:51 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:38:51 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:55 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/04 11:38:55 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:55 [WARN] Test: Step plan: DIFF:

DESTROY: alicloud_cdn_domain.domain

STATE:

alicloud_cdn_domain.domain:
  ID = tf-testacc6068554464636384429.xiaozhu.com
  block_ips.# = 0
  cache_config.# = 0
  cdn_type = web
  certificate_config.# = 1
  certificate_config.3358202545.private_key = -----BEGIN PRIVATE KEY-----
MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQDOdY9ITNHamP4B
3neS4bEm00oV5sNkqoNfGrwaR7qfwvGM76VXXb0VHnROeInXAn3QlYaFhLXDRVyJ
CqVg6SJhbaL6LqXTXQtl6ARxAuNqK/WQPsFbGgoofkkbgd6/sjc1x4fq5lGtBS3o
o/8dIFKL5s51LJO1DDM8CS2cJrsovCwrokyJUM0PUU3WmKNreDrhlZHUIic7GIkU
TTzqgirWL2HBAS2pSxTnFZ9GhvyPRZ9D963PHyvQRB2qGEbp5rrFAnQnCqYrBxym
HbZnd2YM/mfjwZIwEbbnzupDl0qL+ru2sy0MWYGSZnmoDZKkOP0fvW+BkOnK49D4
uueUFmVRAgMBAAECggEAGFrP2zSMsN/JXwkSS/Zpwm28WJcPR6nBs49gzyzU/BGw
EvMWKxc4vewIxlT71axKkTeCVe/QzUc6YkQqPCNkVd/sEN093JAmTxAurfIsR5MF
9c0hXBDXT+2NzDvmvfBVCPgPtYsT6Xgp8T6fUp1Ef5JrmnD2v62/wX5Hrhr3ixdp
FoEJ92/FJN+Dt/Duri481+RYqUUbNqiZAuEMsny/hM+aijtpQgZK8FqnVfjIOo0l
p4TwfDoxtTwkj0QeJxelX/sNSD+iQ5xVmnxQ9Q1qmKMk0M1Wl8318ovcrXDTN3RZ
R/Fp+lNw0aC4OLuOJDc7A13l5unHmex4Ka1kbjOZwQKBgQD8w0hJiVVmiojJ3/YX
06rG3ufGlKXu4qvRL6eXNqnDp8AnKXpjKiDV5lGz+vK0pcAyF5AhpddDN20Un+nR
4VU7NslNJdBfjQxfEKJRkGBP22C+pdlxOBDgdgouN9q7TTr9StaQC0KF+HUcF8fO
sYpEp1R+27iHie+F6uS78ur0mQKBgQDRGnbv8acTDTBW8nll1wTRa0+iVR2yxglS
ndC97HzYHOFb+caVmaM18QzHaaRvAhQ0bZXOjFV0QVv1nytlmHxsMlG/N6LDz5jA
NLGvbrzDzki+RUk0TWJuYpvdevsnwaCxMwyY6CQ5MKhKAvwgE8AhdAq5MGcE+6U/
oAOZNyOxeQKBgQCiCB2i5mLUpSIjJ2r+wzXK3sH9zvTAOpaiNsZcbTJOto67jB9k
ynDaLhdaJRjJLSgT9H700vc3o6RNgGXHoYedufU5e3AkkKrJlkQ3vTHAf4V5MaA+
sA5BlenYzv1s7IlQLlV1aYJvl2Kba7MukSlt8UZ9PCUC3i2pz3Zp9cMgoQKBgQCK
GkR7bMq/1nIausJa9IwGFC3gNP8MV6dInVqEVXCO+2QL7wetPm+A7NdXzPoBJwpZ
JhdO93ho89Hcg2eSDgf/HazH8eLaGH32U9cW2rhpShDZOcGDfaiI5y+yM8s1Erki
z2h+hLOH4g8D8ry6ItE+RvneHY2syNb3EqPNyZEVYQKBgGeC4KiyTpW+SqmkDRU/
/FxIAlZE2vWmtJptAUkmLEVoBi0/GvJoYYJR4Jhj4gS9tJcrI860X0j05NZHvIX8
UJ/vZZeiFRiatXYUtKbD/ngoOXTF92ew2J3eCwNHH5OKwEO3aejaYwD9m0QM4kTl
VLWIf79+G3NrxIeqPnqv+l3J
-----END PRIVATE KEY-----
  certificate_config.3358202545.server_certificate = -----BEGIN CERTIFICATE-----
MIICrDCCAZQCCQDApyXUTYDE+DANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA0q
LnhpYW96aHUuY29tMB4XDTE4MTIzMTEwMDY1OVoXDTE5MTIzMTEwMDY1OVowGDEW
MBQGA1UEAwwNKi54aWFvemh1LmNvbTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCC
AQoCggEBAM51j0hM0dqY/gHed5LhsSbTShXmw2Sqg18avBpHup/C8YzvpVddvRUe
dE54idcCfdCVhoWEtcNFXIkKpWDpImFtovoupdNdC2XoBHEC42or9ZA+wVsaCih+
SRuB3r+yNzXHh+rmUa0FLeij/x0gUovmznUsk7UMMzwJLZwmuyi8LCuiTIlQzQ9R
TdaYo2t4OuGVkdQiJzsYiRRNPOqCKtYvYcEBLalLFOcVn0aG/I9Fn0P3rc8fK9BE
HaoYRunmusUCdCcKpisHHKYdtmd3Zgz+Z+PBkjARtufO6kOXSov6u7azLQxZgZJm
eagNkqQ4/R+9b4GQ6crj0Pi655QWZVECAwEAATANBgkqhkiG9w0BAQsFAAOCAQEA
JAWcQb7942jWDWjFqW5C4eyJBuJxK8dsTsYpOa0ccpp/A2cVCTNMzV/sMCGROy2B
uhrTMG4q3QawFwV+K4nQX8yJUJp57zyUfdit+yOD4hDzPyF5MOlOwzdqeg9Y9ODL
Sc2O6J3tFsx2332Hb7RXFlYodEk1uFezrj4YLVrAgk1QojaBHpuFiA/O3eCmErjW
c88bcil60qMvlhBhCWaZzivji6oc3y/6qWRZ3apxp9/6sjOvlm5Q+wwoLHXaU+L6
w7TZptw6upuSrAS+N4Rwqgn5TfPvbGkdRG0X0TDzE1+F+w377kE71nV8lmi8F786
bqloD1gWJER071WRz1coHQ==
-----END CERTIFICATE-----

  certificate_config.3358202545.server_certificate_status = on
  domain_name = tf-testacc6068554464636384429.xiaozhu.com
  http_header_config.# = 0
  optimize_enable = 
  page_compress_enable = 
  range_enable = 
  scope = overseas
  source_port = 80
  source_type = oss
  sources.# = 1
  sources.21942838 = terraformtest.aliyuncs.com
  video_seek_enable =
2019/01/04 11:38:55 [INFO] terraform: building graph: GraphTypeApply
2019/01/04 11:38:55 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:59 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/04 11:38:59 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:59 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/04 11:38:59 [WARN] terraform: shadow graph disabled
2019/01/04 11:38:59 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/04 11:38:59 [WARN] terraform: shadow graph disabled
--- PASS: TestAccAlicloudCdnDomain_https (412.07s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	412.119s
```

xiaozhu36 replied:
OK. Please check other comments.

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -98,29 +98,34 @@ The config supports the following:
+    * `cache_content` - (Required) Cache content of the cache config. It's value is a path string when the `cache_type` is `path`. When the `cache_type` is `suffix`, it's value is a string which contains multiple file suffixes separated by commas.
     * `ttl` - (Required, Type: int) Cache time of the cache config.
     * `weight` - (Optional, Type: int) Weight of the cache config. This parameter's value is between 1 and 99. Default value is `1`. The higher the value, the higher the priority.
 
```
Can you improve this docs by using `block` to arrange several configs?

chanind replied:
I'm sorry I don't understand, what does it mean to use `block` to arrange configs? Is there an example in another part of the docs I can use as a reference? Sorry, I'm new to this

xiaozhu36 replied:
Please refer to: https://github.com/terraform-providers/terraform-provider-alicloud/blob/master/website/docs/r/oss_bucket.html.markdown#block-website

chanind commented on alicloud/resource_alicloud_cdn_domain.go:
```diff
@@ -716,6 +770,65 @@ func authConfigUpdate(client *connectivity.AliyunClient, d *schema.ResourceData)
+
+	val := newConfig.List()[0].(map[string]interface{})
+	d.SetPartial("certificate_config")
+	args.ServerCertificateStatus = val["server_certificate_status"].(string)
```
done in e038fe4

xiaozhu36 replied:
I think line 797 should be removed.

chanind replied:
@xiaozhu36 you need to submit a private key as part of the API request to set up SSL. If I delete line 797 then there is no way for SSL to be set up. SSL requires a private key and a certificate.

chanind commented:
@xiaozhu36 I updated the PR and changed to use `TypeList` instead of `TypeSet` like you suggested, so the test doesn't have a hash anymore. Here's the updated test results:

```
=== RUN   TestAccAlicloudCdnDomain_https
2019/01/08 14:14:20 [WARN] Test: Executing step 0
2019/01/08 14:14:20 [INFO] terraform: building graph: GraphTypeValidate
2019/01/08 14:14:20 [WARN] terraform: shadow graph disabled
2019/01/08 14:14:20 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/08 14:14:20 [WARN] terraform: shadow graph disabled
2019/01/08 14:14:20 [INFO] terraform: building graph: GraphTypePlan
2019/01/08 14:14:20 [WARN] terraform: shadow graph disabled
2019/01/08 14:14:20 [WARN] Test: Step plan: DIFF:

CREATE: alicloud_cdn_domain.domain
  cdn_type:                                       "" => "web"
  certificate_config.#:                           "0" => "1"
  certificate_config.0.private_key:               "<sensitive>" => "<sensitive>" (attribute changed)
  certificate_config.0.server_certificate:        "<sensitive>" => "<sensitive>" (attribute changed)
  certificate_config.0.server_certificate_status: "" => "on"
  domain_name:                                    "" => "tf-testacc5760878342779004972.xiaozhu.com"
  scope:                                          "" => "overseas"
  source_port:                                    "" => "80"
  source_type:                                    "" => "oss"
  sources.#:                                      "0" => "1"
  sources.21942838:                               "" => "terraformtest.aliyuncs.com"

STATE:

<no state>
2019/01/08 14:14:20 [INFO] terraform: building graph: GraphTypeApply
2019/01/08 14:14:20 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:05 [WARN] Domain id "tf-testacc5760878342779004972.xiaozhu.com"
2019/01/08 14:21:05 [INFO] terraform: building graph: GraphTypePlan
2019/01/08 14:21:05 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:05 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/08 14:21:05 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:13 [INFO] terraform: building graph: GraphTypePlan
2019/01/08 14:21:13 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:13 [WARN] Test: Running ID-only refresh check on tf-testacc5760878342779004972.xiaozhu.com
2019/01/08 14:21:13 [WARN] Test: Executing destroy step
2019/01/08 14:21:13 [INFO] terraform: building graph: GraphTypeValidate
2019/01/08 14:21:13 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:13 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/08 14:21:13 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:19 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/08 14:21:19 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:19 [WARN] Test: Step plan: DIFF:

DESTROY: alicloud_cdn_domain.domain

STATE:

alicloud_cdn_domain.domain:
  ID = tf-testacc5760878342779004972.xiaozhu.com
  block_ips.# = 0
  cache_config.# = 0
  cdn_type = web
  certificate_config.# = 1
  certificate_config.0.private_key = -----BEGIN PRIVATE KEY-----
MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQDOdY9ITNHamP4B
3neS4bEm00oV5sNkqoNfGrwaR7qfwvGM76VXXb0VHnROeInXAn3QlYaFhLXDRVyJ
CqVg6SJhbaL6LqXTXQtl6ARxAuNqK/WQPsFbGgoofkkbgd6/sjc1x4fq5lGtBS3o
o/8dIFKL5s51LJO1DDM8CS2cJrsovCwrokyJUM0PUU3WmKNreDrhlZHUIic7GIkU
TTzqgirWL2HBAS2pSxTnFZ9GhvyPRZ9D963PHyvQRB2qGEbp5rrFAnQnCqYrBxym
HbZnd2YM/mfjwZIwEbbnzupDl0qL+ru2sy0MWYGSZnmoDZKkOP0fvW+BkOnK49D4
uueUFmVRAgMBAAECggEAGFrP2zSMsN/JXwkSS/Zpwm28WJcPR6nBs49gzyzU/BGw
EvMWKxc4vewIxlT71axKkTeCVe/QzUc6YkQqPCNkVd/sEN093JAmTxAurfIsR5MF
9c0hXBDXT+2NzDvmvfBVCPgPtYsT6Xgp8T6fUp1Ef5JrmnD2v62/wX5Hrhr3ixdp
FoEJ92/FJN+Dt/Duri481+RYqUUbNqiZAuEMsny/hM+aijtpQgZK8FqnVfjIOo0l
p4TwfDoxtTwkj0QeJxelX/sNSD+iQ5xVmnxQ9Q1qmKMk0M1Wl8318ovcrXDTN3RZ
R/Fp+lNw0aC4OLuOJDc7A13l5unHmex4Ka1kbjOZwQKBgQD8w0hJiVVmiojJ3/YX
06rG3ufGlKXu4qvRL6eXNqnDp8AnKXpjKiDV5lGz+vK0pcAyF5AhpddDN20Un+nR
4VU7NslNJdBfjQxfEKJRkGBP22C+pdlxOBDgdgouN9q7TTr9StaQC0KF+HUcF8fO
sYpEp1R+27iHie+F6uS78ur0mQKBgQDRGnbv8acTDTBW8nll1wTRa0+iVR2yxglS
ndC97HzYHOFb+caVmaM18QzHaaRvAhQ0bZXOjFV0QVv1nytlmHxsMlG/N6LDz5jA
NLGvbrzDzki+RUk0TWJuYpvdevsnwaCxMwyY6CQ5MKhKAvwgE8AhdAq5MGcE+6U/
oAOZNyOxeQKBgQCiCB2i5mLUpSIjJ2r+wzXK3sH9zvTAOpaiNsZcbTJOto67jB9k
ynDaLhdaJRjJLSgT9H700vc3o6RNgGXHoYedufU5e3AkkKrJlkQ3vTHAf4V5MaA+
sA5BlenYzv1s7IlQLlV1aYJvl2Kba7MukSlt8UZ9PCUC3i2pz3Zp9cMgoQKBgQCK
GkR7bMq/1nIausJa9IwGFC3gNP8MV6dInVqEVXCO+2QL7wetPm+A7NdXzPoBJwpZ
JhdO93ho89Hcg2eSDgf/HazH8eLaGH32U9cW2rhpShDZOcGDfaiI5y+yM8s1Erki
z2h+hLOH4g8D8ry6ItE+RvneHY2syNb3EqPNyZEVYQKBgGeC4KiyTpW+SqmkDRU/
/FxIAlZE2vWmtJptAUkmLEVoBi0/GvJoYYJR4Jhj4gS9tJcrI860X0j05NZHvIX8
UJ/vZZeiFRiatXYUtKbD/ngoOXTF92ew2J3eCwNHH5OKwEO3aejaYwD9m0QM4kTl
VLWIf79+G3NrxIeqPnqv+l3J
-----END PRIVATE KEY-----
  certificate_config.0.server_certificate = -----BEGIN CERTIFICATE-----
MIICrDCCAZQCCQDApyXUTYDE+DANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA0q
LnhpYW96aHUuY29tMB4XDTE4MTIzMTEwMDY1OVoXDTE5MTIzMTEwMDY1OVowGDEW
MBQGA1UEAwwNKi54aWFvemh1LmNvbTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCC
AQoCggEBAM51j0hM0dqY/gHed5LhsSbTShXmw2Sqg18avBpHup/C8YzvpVddvRUe
dE54idcCfdCVhoWEtcNFXIkKpWDpImFtovoupdNdC2XoBHEC42or9ZA+wVsaCih+
SRuB3r+yNzXHh+rmUa0FLeij/x0gUovmznUsk7UMMzwJLZwmuyi8LCuiTIlQzQ9R
TdaYo2t4OuGVkdQiJzsYiRRNPOqCKtYvYcEBLalLFOcVn0aG/I9Fn0P3rc8fK9BE
HaoYRunmusUCdCcKpisHHKYdtmd3Zgz+Z+PBkjARtufO6kOXSov6u7azLQxZgZJm
eagNkqQ4/R+9b4GQ6crj0Pi655QWZVECAwEAATANBgkqhkiG9w0BAQsFAAOCAQEA
JAWcQb7942jWDWjFqW5C4eyJBuJxK8dsTsYpOa0ccpp/A2cVCTNMzV/sMCGROy2B
uhrTMG4q3QawFwV+K4nQX8yJUJp57zyUfdit+yOD4hDzPyF5MOlOwzdqeg9Y9ODL
Sc2O6J3tFsx2332Hb7RXFlYodEk1uFezrj4YLVrAgk1QojaBHpuFiA/O3eCmErjW
c88bcil60qMvlhBhCWaZzivji6oc3y/6qWRZ3apxp9/6sjOvlm5Q+wwoLHXaU+L6
w7TZptw6upuSrAS+N4Rwqgn5TfPvbGkdRG0X0TDzE1+F+w377kE71nV8lmi8F786
bqloD1gWJER071WRz1coHQ==
-----END CERTIFICATE-----

  certificate_config.0.server_certificate_status = on
  domain_name = tf-testacc5760878342779004972.xiaozhu.com
  http_header_config.# = 0
  optimize_enable = 
  page_compress_enable = 
  range_enable = 
  scope = overseas
  source_port = 80
  source_type = oss
  sources.# = 1
  sources.21942838 = terraformtest.aliyuncs.com
  video_seek_enable =
2019/01/08 14:21:19 [INFO] terraform: building graph: GraphTypeApply
2019/01/08 14:21:19 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:24 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/08 14:21:24 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:24 [INFO] terraform: building graph: GraphTypeRefresh
2019/01/08 14:21:24 [WARN] terraform: shadow graph disabled
2019/01/08 14:21:24 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/01/08 14:21:24 [WARN] terraform: shadow graph disabled
--- PASS: TestAccAlicloudCdnDomain_https (424.22s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	424.274s
```

xiaozhu36 commented:
Hi @chanind Can you rebase this PR before the next submitting?

chanind commented:
@xiaozhu36 I deleted line 797 and rebased / squashed

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -95,34 +95,46 @@ The config supports the following:
-    * `ttl` - (Required, Type: int) Cache time of the cache config.
-    * `weight` - (Optional, Type: int) Weight of the cache config. This parameter's value is between 1 and 99. Default value is `1`. The higher the value, the higher the priority.
+### Block parameter_filter_config
+`parameter_filter_config` - (Optional, Type: set) Parameter filter config of the accelerated domain. It's a set and consists of at most one item.
```
Please add a blank line between line 98 and 99

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -95,34 +95,46 @@ The config supports the following:
+* `hash_key_args` - (Optional, Type: list) Reserved parameters of `parameter_filter_config`. It's a list of string and consists of at most 10 items.
+
+### Block page_404_config
+`page_404_config` - (Optional, Type: set) Error Page config of the accelerated domain. It's a set and consists of at most one item.
```
Please add a blank line between line 103 and 104

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -95,34 +95,46 @@ The config supports the following:
+* `custom_page_url` - (Optional) Custom page url of the error page. It must be the full path under the accelerated domain name. It's value must be `http://promotion.alicdn.com/help/oss/error.html` when `page_type` value is `charity` and It can not be set when `page_type` value is `default`.
+
+### Block refer_config
+`refer_config` - (Optional, Type: set) Refer anti-theft chain config of the accelerated domain. It's a set and consists of at most 1 item.
```
Please add a blank line between line 108 and 109

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -95,34 +95,46 @@ The config supports the following:
+
+### Block auth_config
+`auth_config` - (Optional, Type: set)  Auth config of the accelerated domain. It's a set and consist of at most 1 item.
+* `auth_type` - (Optional) Auth type of the auth config. Valid values are  `no_auth`, `type_a`, `type_b` and `type_c`. Default value is `no_auth`.
```
Please add a blank line between line 114 and 115

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -95,34 +95,46 @@ The config supports the following:
+* `timeout` - (Optional, Type: int)  Authentication cache time of the auth config. Default value is `1800`. It's value is valid only when the `auth_type` is `type_b` or `type_c`.
+
+### Block certificate_config
+`certificate_config` - (Optional, Type: set)  Certificate config of the accelerated domain. It's a set and consist of at most 1 item.
```
Please add a blank line between line 121 and 122

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -95,34 +95,46 @@ The config supports the following:
+* `private_key` - (Optional) The SSL private key. This is required if `server_certificate_status` is `on`
+
+### Block http_header_config
+`http_header_config` - (Optional, Type: set) Http header config of the accelerated domain. It's a set and consist of at most 8 items. The `header_key` for each item can not be repeated.
```
Please add a blank line between line 127 and128

xiaozhu36 commented on website/docs/r/cdn_domain.html.markdown:
```diff
@@ -95,34 +95,46 @@ The config supports the following:
+* `header_value` - (Required) Header value of the http header.
+
+### Block cache_config
+`cache_config` - (Optional, Type: set)  Cache config of the accelerated domain. It's a set and each item's `cache_content` can not be repeated.
```
Please add a blank line between line 132 and 133

chanind commented:
@xiaozhu36 blank lines have been added in bbe7514
